engine/src/main/java/org/terasology/rendering/nui/widgets/models/Tree.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.nui.widgets.models;
+
+import com.google.common.collect.Lists;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * A general-purpose tree data structure.
+ * <p>
+ * Stores a single object. May have a single parent and an arbitrary amount of children.
+ *
+ * @param <T> Type of objects stored in the tree.
+ */
+public class Tree<T> {
+    private static final String NULL_NODE_ARGUMENT = "node argument is null";
+    private static final String NODE_ARGUMENT_INVALID_PARENT = "node argument is not a child of this tree";
+    private static final String ITERATOR_NO_ITEMS = "no elements left (try validating with hasNext?)";
+
+    /**
+     * The object stored in this tree.
+     */
+    private T value;
+    /**
+     * The parent node for this tree.
+     */
+    private Tree<T> parent;
+    /**
+     * The list of children for this tree.
+     */
+    private List<Tree<T>> children = Lists.newArrayList();
+    /**
+     * Whether the tree is expanded in the interface.
+     */
+    private boolean expanded;
+
+    public Tree() {
+        this(null);
+    }
+
+    public Tree(T value) {
+        this.value = value;
+    }
+
+    /**
+     * @return A shallow copy of this tree.
+     */
+    public Tree<T> copy() {
+        Tree<T> copy = new Tree<>(this.value);
+        copy.setExpanded(this.expanded);
+
+        for (Tree<T> child : this.children) {
+            copy.addChild(child.copy());
+        }
+        return copy;
+    }
+
+    /**
+     * @return This tree's parent.
+     */
+    public Tree<T> getParent() {
+        return this.parent;
+    }
+
+    /**
+     * @return Whether the tree is a root (has no parent node).
+     */
+    public boolean isRoot() {
+        return this.parent == null;
+    }
+
+    /**
+     * @return The list of children for this tree.
+     */
+    public List<Tree<T>> getChildren() {
+        return this.children;
+    }
+
+
+    /**
+     * @param node The {@code Tree}the index of which is to be returned.
+     * @return The index of the specified {@code Tree}.
+     */
+    public int getIndex(Tree<T> node) {
+        if (node == null) {
```

Comment:
Just noticed the notification for this comment. Fixed all the checks accordingly.